### ISSUE TITLE
trivial unification problem arbitrarily slow with primitive projections

### ISSUE THREAD

**Issue by andres-erbsen** [coq/coq#9573]
#### Description of the problem

When unifying terms with primitive projections, the heuristics choose to inline unrelated lets in the proof context before evaluating a projection from a record at the only location where lhs and rhs differ.

```coq
Set Primitive Projections.
Require Import Coq.ZArith.BinIntDef.
Open Scope Z_scope.
Record word := { t : Type; get : t -> Z; add : t -> t -> t }.
Record parameters := { pword : word }.
Record nt := mk { nU : Z }.
Definition nW : word := {| t := nt; get := nU; add x y := mk (Z.add (nU x) (nU y)) |}.
Definition p : parameters := {| pword := nW |}.
Goal forall x : t nW,
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     (get nW x = 0) = (get (pword p) x = 0).
Proof.
  intros.
  timeout 1 assert (nW = (pword p)) as _ by reflexivity.
  timeout 1 assert ((get nW x = 0) = (get (pword p) x = 0)) as _ by (cbv [pword]; reflexivity).
  Timeout 1 reflexivity.
  (* Timeout! *)
Qed.
```

The timeout does not occur without `Set Primitive Projections`.

cc @ppedrot @mattam82 
thanks @JasonGross for helping me minimize this
@samuelgruetter this is what is making bedrock2 word slow for me (in range propagation)

#### Coq Version

8.9.0

**Comment by JasonGross** [comment]
Here is a slightly smaller example:
```coq
Set Primitive Projections.
Require Import Coq.ZArith.BinIntDef.
Definition nU (x : Z) : Z := x.
Record word := { get : Z -> Z; add : Z -> Z -> Z }.
Definition nW : word := {| get x := x; add x y := Z.add (nU x) (nU y) |}.
Definition nW' := nW.
Goal forall x : Z,
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     let x := add nW x x in
     (get nW x) = (get nW' x).
Proof.
  intros.
  Timeout 1 Time symmetry; reflexivity. (* Finished transaction in 0. secs (0.u,0.s) (successful) *)
  Undo.
  Time reflexivity. (* Finished transaction in 3.121 secs (3.125u,0.s) (successful) *)
  Undo.
  Strategy -1000 [nW'].
  Time reflexivity. (* Finished transaction in 0. secs (0.u,0.s) (successful) *)
```

My guess at what is happening is that Coq notices that `get nW x` is not syntactically the same as `get nW' x`, decides to unfold the LHS before the RHS, decides to unfold `nW` first, then inlines `get nW` so that we get `x`, then unfolds `x` and continues to unfold it until it is fully inlined; since things still don't unify it then unfolds `nW'`, then `nW`, and then inlines `x` on the RHS.  Since the conversion algorithm (apparently) biases unfolding on the LHS, using `symmetry` makes it instantaneous.  Can a unification expert (@herbelin ?) confirm that this is what is going on?  (Neither `Set Debug Unification.` nor `Set Debug RAKAM.` give me anything useful; perhaps there should be a more verbose unification debug setting?)

I'm a bit baffled why removing `nU` in the `Z.add` definitions makes things go through more quickly, though...... perhaps `nU` is unfolded on the LHS and this somehow forces the inlining or something?  Can someone explain this to me?

@andres-erbsen This suggests that `Strategy -100 [pword p].` is a workaround for your issue.

**Comment by mattam82** [comment]
@JasonGross does your example change when you unset primitive projections? I think your analysis is correct, as soon as conversion choses to unfold nW first you’re toast.

**Comment by mattam82** [comment]
Maybe what we’re missing is the first order approximation heuristic for primitive projection applications, indeed it would seem natural to not try to unfold x at all here, which appears to be the case given the time spent in the second test.

**Comment by JasonGross** [comment]
> does your example change when you unset primitive projections

Yes, unsetting primitive projections fixes the performance issue

> Maybe what we’re missing is the first order approximation heuristic for primitive projection applications

If that is the difference between primitive projections and not, then, yes, I think it should be added.

**Comment by maximedenes** [comment]
@mattam82 can I assign this issue to you?
